<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Use example - AsaruSim</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Use example";
        var mkdocs_page_input_path = "examples.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> AsaruSim
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../introduction/">Introduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../parameters/">Input parameters</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Use example</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#examples">EXAMPLES</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#sample-data">Sample data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#basic-workflow">BASIC WORKFLOW</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#with-pcr-amplifiction">WITH PCR AMPLIFICTION</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#with-simulated-cell-type-counts">WITH SIMULATED CELL TYPE COUNTS</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#with-personalized-error-model">WITH PERSONALIZED ERROR MODEL</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#complete-workflow">COMPLETE WORKFLOW</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">AsaruSim</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Use example</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="usage">Usage</h2>
<p>User can choose among 4 ways to simulate template reads.
- use a real count matrix
- estimated the parameter from a real count matrix to simulate synthetic count matrix 
- specified by his/her own the input parameter
- a combination of the above options</p>
<p>We use SPARSIM tools to simulate count matrix. for more information a bout synthetic count matrix, please read <a href="https://gitlab.com/sysbiobig/sparsim/-/blob/master/vignettes/sparsim.Rmd?ref_type=heads#Sec_Input_parameter_estimated_from_data">SPARSIM</a> documentaion.</p>
<h3 id="examples">EXAMPLES</h3>
<h5 id="sample-data">Sample data</h5>
<p>A demonstration dataset to initiate this workflow is accessible on zenodo DOI : <a href="https://zenodo.org/records/12731409">10.5281/zenodo.12731408</a>. This dataset is a subsample from a Nanopore run of the <a href="https://www.10xgenomics.com/datasets/5k-human-pbmcs-3-v3-1-chromium-controller-3-1-standard">10X 5k human pbmcs</a>.</p>
<p>The human GRCh38 <a href="https://ftp.ensembl.org/pub/release-112/fasta/homo_sapiens/cdna/">reference transcriptome</a>, <a href="https://ftp.ensembl.org/pub/release-112/gtf/homo_sapiens/">gtf annotation</a> and <a href="https://ftp.ensembl.org/pub/release-112/fasta/homo_sapiens/dna/">fasta referance genome</a> can be downloaded from Ensembl.</p>
<h5 id="basic-workflow">BASIC WORKFLOW</h5>
<pre><code class="language-bash"> nextflow run main.nf --matrix dataset/sub_pbmc_matrice.csv \
                      --transcriptome dataset/Homo_sapiens.GRCh38.cdna.all.fa \
                      --features gene_name \
                      --gtf dataset/genes.gtf
</code></pre>
<h5 id="with-pcr-amplifiction">WITH PCR AMPLIFICTION</h5>
<pre><code class="language-bash"> nextflow run main.nf --matrix dataset/sub_pbmc_matrice.csv \
                      --transcriptome dataset/Homo_sapiens.GRCh38.cdna.all.fa \
                      --features gene_name \
                      --gtf dataset/GRCh38-2020-A-genes.gtf \
                      --pcr_cycles 2 \
                      --pcr_dup_rate 0.7 \
                      --pcr_error_rate 0.00003
</code></pre>
<h5 id="with-simulated-cell-type-counts">WITH SIMULATED CELL TYPE COUNTS</h5>
<pre><code class="language-bash"> nextflow run main.nf --matrix dataset/sub_pbmc_matrice.csv \
                      --transcriptome dataset/Homo_sapiens.GRCh38.cdna.all.fa \
                      --features gene_name \
                      --gtf dataset/GRCh38-2020-A-genes.gtf \
                      --sim_celltypes true \
                      --cell_types_annotation dataset/sub_pbmc_cell_type.csv
</code></pre>
<h5 id="with-personalized-error-model">WITH PERSONALIZED ERROR MODEL</h5>
<pre><code class="language-bash">nextflow run main.nf --matrix dataset/sub_pbmc_matrice.csv \
                     --transcriptome dataset/Homo_sapiens.GRCh38.cdna.all.fa \
                     --features gene_name \
                     --gtf dataset/GRCh38-2020-A-genes.gtf \
                     --build_model true \
                     --fastq_model dataset/sub_pbmc_reads.fq \
                     --ref_genome dataset/GRCh38-2020-A-genome.fa 
</code></pre>
<h5 id="complete-workflow">COMPLETE WORKFLOW</h5>
<pre><code class="language-bash"> nextflow run main.nf --matrix dataset/sub_pbmc_matrice.csv \
                      --transcriptome dataset/Homo_sapiens.GRCh38.cdna.all.fa \
                      --features gene_name \
                      --gtf dataset/GRCh38-2020-A-genes.gtf \
                      --sim_celltypes true \
                      --cell_types_annotation dataset/sub_pbmc_cell_type.csv
                      --build_model true \
                      --fastq_model dataset/sub_pbmc_reads.fq \
                      --ref_genome dataset/GRCh38-2020-A-genome.fa 
                      --pcr_cycles 2 \
                      --pcr_dup_rate 0.7 \
                      --pcr_error_rate 0.00003
</code></pre>
<h2 id="results">Results</h2>
<p>After execution, results will be available in the specified <code>--outdir</code>. This includes simulated Nanopore reads <code>.fastq</code>, along with log files and QC report.</p>
<h2 id="cleaning-up">Cleaning Up</h2>
<p>To clean up temporary files generated by Nextflow:</p>
<pre><code class="language-bash">nextflow clean -f
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../parameters/" class="btn btn-neutral float-left" title="Input parameters"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../parameters/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
